tracee-ebpf/main.go

```diff
@@ -382,58 +408,62 @@ func prepareOutput(outputSlice []string, containerMode bool) (tracee.OutputConfi
 		case "option":
 			switch outputParts[1] {
 			case "stack-addresses":
-				res.StackAddresses = true
+				outcfg.StackAddresses = true
 			case "detect-syscall":
-				res.DetectSyscall = true
+				outcfg.DetectSyscall = true
 			case "exec-env":
-				res.ExecEnv = true
+				outcfg.ExecEnv = true
 			case "relative-time":
-				res.RelativeTime = true
+				outcfg.RelativeTime = true
 			default:
-				return res, nil, fmt.Errorf("invalid output option: %s, use '--output help' for more info", outputParts[1])
+				return outcfg, printcfg, fmt.Errorf("invalid output option: %s, use '--output help' for more info", outputParts[1])
 			}
 		default:
-			return res, nil, fmt.Errorf("invalid output value: %s, use '--output help' for more info", outputParts[1])
+			return outcfg, printcfg, fmt.Errorf("invalid output value: %s, use '--output help' for more info", outputParts[1])
 		}
 	}
 
-	outf := os.Stdout
-	if outPath != "" {
-		// To avoid accidental deletions, make sure this is not a directory
+	printcfg.kind = printerKind
+
+	if outPath == "" {
+		printcfg.outFile = os.Stdout
+	} else {
+		printcfg.outPath = outPath
 		fileInfo, err := os.Stat(outPath)
-		if err == nil && fileInfo.IsDir() {
-			return res, nil, fmt.Errorf("cannot use a path of existing directory %s", outPath)
-		}
-		dir := filepath.Dir(outPath)
-		os.MkdirAll(dir, 0755)
-		os.Remove(outPath)
-		outf, err = os.Create(outPath)
-		if err != nil {
-			return res, nil, fmt.Errorf("failed to create output path: %v", err)
+		if err == nil {
+			if fileInfo.IsDir() {
+				return outcfg, printcfg, fmt.Errorf("cannot use a path of existing directory %s", outPath)
+			}
+		} else {
+			dir := filepath.Dir(outPath)
+			os.MkdirAll(dir, 0755)
+			printcfg.outFile, err = os.Create(outPath)
+			if err != nil {
+				return outcfg, printcfg, fmt.Errorf("failed to create output path: %v", err)
+			}
 		}
 	}
-	errf := os.Stderr
-	if errPath != "" {
-		// To avoid accidental deletions, make sure this is not a directory
+
+	if errPath == "" {
+		printcfg.errFile = os.Stderr
+	} else {
+		printcfg.errPath = errPath
 		fileInfo, err := os.Stat(errPath)
-		if err == nil && fileInfo.IsDir() {
-			return res, nil, fmt.Errorf("cannot use a path of existing directory %s", errPath)
-		}
-		dir := filepath.Dir(errPath)
-		os.MkdirAll(dir, 0755)
-		os.Remove(errPath)
-		errf, err = os.Create(errPath)
-		if err != nil {
-			return res, nil, fmt.Errorf("failed to create error path: %v", err)
+		if err == nil {
+			if fileInfo.IsDir() {
+				return outcfg, printcfg, fmt.Errorf("cannot use a path of existing directory %s", errPath)
+			}
+		} else {
+			dir := filepath.Dir(errPath)
+			os.MkdirAll(dir, 0755)
+			printcfg.errFile, err = os.Create(errPath)
+			if err != nil {
+				return outcfg, printcfg, fmt.Errorf("failed to create output path: %v", err)
+			}
 		}
 	}
```

Comment:
If you already took newEventPrinter out of this function, I don't see a reason to keep all of the oufFile and errFile logic here - we can move it to a different function (preparePrinter or something like that)

Reply:
some `--output` options will be relevant for both OutputConfig and printerConfig. if we split parsing into 2 functions like you suggest we would be parsing the same cli options in different places. I don't mind doing it if you think it's better, just raising this concern.

Reply:
I don't see any line above (lines 426-462) that changes something in outcfg, but only printcfg.
What I suggest is to split into two functions (I don't mind about the names, as long as they are meaningful):
1. Looping the user input and returning outcfg, outPath and errPath
2. Using outPath and errPath for the creation of ourFile and errFile used by the printer

Reply:
> Looping the user input and returning outcfg, outPath and errPath

and also `printerKind`. 

So `prepareOutput` ended up returning: `outcfg, outpath, errpath, printerkind`, which is basically all the ingredients for `printerConfig`. It might as well just return the printerconfig already, no? If you still think it's valid, I can make the change

Reply:
I've implemented (I think) what you asked in this commit, tell me if this is what you want, and if so I'll add to the PR https://github.com/itaysk/tracee/commit/124996f4011f22a153394e1ac514b78ef619e12f

Reply:
> So prepareOutput ended up returning: outcfg, outpath, errpath, printerkind, which is basically all the ingredients for printerConfig. It might as well just return the printerconfig already, no? If you still think it's valid, I can make the change

This is the situation today - it returns eventPrinter and OutputConfig, but you changed it in your PR.

I think that there are two things in this PR that we can split:
1. resolving "wait for tracee-ebpf to load"
2. Splitting preparePrinter and prepareOutput

If possible, I would avoid having (2) in this PR

Reply:
it's in this PR for a reason, see #3 in the PR description

Reply:
Ok, so please push the latest change so we can review it here